plenum/server/consensus/consensus_data_provider.py

```diff
@@ -11,12 +15,17 @@ class ConsensusDataProvider:
     TODO: Consider adding persistent local storage for 3PC certificates
     TODO: Restore primary name from audit ledger instead of passing through constructor
     """
-    def __init__(self, name: str, validators: List[str], primary_name: str):
+
+    def __init__(self, name: str):
         self._name = name
-        self.set_validators(validators)
         self.view_no = 0
+        self.pp_seq_no = 0
         self.waiting_for_new_view = False
-        self.primary_name = primary_name
+        self.primary_name = None
+        self._stable_checkpoint = None
+        self._checkpoints = SortedDict(lambda k: k[1])
```

Comment:
It would be nice to add comment describing mapping here. If you just want to store sorted checkpoint list you can use SortedList or SortedListWithKey

Reply:
I've decided to use dict, cause in replica code we do not store checkpoints, but only CheckpointState, that identified by (start_seq_no, end_seq_no)

Reply:
The problem here is that services that are going to use this (like ViewChangeService) don't care about neither order nor start_seq_no. All they need to know is a list of currently reached checkpoint which are not yet stabilized. I suspect that you was trying to mimic Replica internal data structures as close as possible in order to minimize additional computations, however I don't think it is really worth optimizing here, especially this way. Most of times this list will contain 0 to 1 elements, and even if it starts accumulating it is unlikely that it will ever grow beyond 10 elemnts.